File: miniBPE/basic.py
import json
from tqdm import tqdm
from collections import Counter

class BasicTokenizer:
  def __init__(self):
    self.vocab_size = 0
    self.vocab = {}
    self.merges = {}

  def _get_stats(self, ids): 
    """
    Takes a list of integers and returns a dictionary of counts of pairs(consecutive ones).
    Eg: [1, 2, 3, 1, 2] -> {(1, 2): 2, (2, 3): 1, (3, 1): 1}
    """
    pairs = [(ids[i], ids[i+1]) for i in range(len(ids) - 1)]
    return Counter(pairs)

  def _merge(self, ids, pair, idx):
    """
    Replaces all consecutive pair with the new integer token idx in the list of integers.
    Eg: ids=[1, 2, 3, 1, 2], pair=(1, 2), idx=4 -> [4, 3, 4]
    """
    return [idx if (ids[i] == pair[0] and ids[i+1] == pair[1]) else ids[i] for i in range(len(ids)-1)]

  def _build_vocab(self, merges):
    """
    Builds the new vocab by merging the merge values together.
    """
    vocab = {idx: bytes([idx]) for idx in range(256)}
    for (p0, p1), idx in merges.items():
      vocab[idx] = vocab[p0] + vocab[p1]
    return vocab

  def train(self, train_data, target_vocab):
    assert target_vocab >= 256
    n_merges = target_vocab - 256
    text_bytes = train_data.encode('utf-8')
    ids = list(text_bytes)

    stats = self._get_stats(ids)
    merges = {}
    for i in tqdm(range(n_merges), desc='Training the tokenizer\t'):
      pair = max(stats, key=stats.get)
      idx = 256 + i
      stats = self._update_stats(stats, ids, pair, idx)
      merges[pair] = idx
    
    vocab = self._build_vocab(merges)
    self.vocab = vocab
    self.merges = merges
    self.vocab_size = len(self.vocab)

    return self.vocab, self.merges

  def _update_stats(self, stats, ids, pair, idx):
    """
    Updates the statistics after each merge.
    """
    new_stats = stats.copy()
    new_stats.pop(pair)
    for i in range(len(ids) - 1):
      if (ids[i], ids[i+1]) == pair:
        new_pair = (idx, ids[i+2] if i+2 < len(ids) else ids[i+1])
        new_stats[new_pair] = new_stats.get(new_pair, 0) + 1
    return new_stats

  def encode(self, text):
    """
    Encodes the input string using 'utf-8' encodings.
    """
    text_bytes = text.encode('utf-8')
    ids = list(text_bytes)
    stats = self._get_stats(ids)
    while len(ids) >= 2:
      pair = min(stats, key=lambda p: self.merges.get(p, float('inf')))
      if pair not in self.merges:
        break

      idx = self.merges[pair]
      stats = self._update_stats(stats, ids, pair, idx)
      ids = [idx if (ids[i] == pair[0] and ids[i+1] == pair[1]) else ids[i] for i in range(len(ids)-1)]
    return ids

  def decode(self, ids):
    """
    Decodes the input list of integers into string.
    """
    text_bytes = b"".join(self.vocab[idx] for idx in ids)
    text = text_bytes.decode('utf-8', errors='replace')
    return text

  def save_model(self, prefix):
    """
    Saves the model to files '.model' & 'vocab.json'.
    """
    model_file = prefix + '.model'
    with open(model_file, 'w') as f:
      for idx1, idx2 in self.merges:
        f.write(f"{idx1} {idx2}\n")
    f.close()

    vocab_file = prefix + '_vocab.json'
    with open(vocab_file, 'w') as f:
      serializable_vocab = {}
      for idx in self.vocab:
        try: 
          serializable_vocab[idx] = str(self.vocab[idx], 'utf-8')
        except:
          pass
      json.dump(serializable_vocab)
    f.close()

  def load_model(self, model_file):
    """
    Loads the model from '.model' file.
    """
    assert model_file.endswith('.model')

    merges = {}
    idx = 256
    with open(model_file, 'r', encoding='utf-8') as f:
      for line in f:
        idx1, idx2 = map(int, line.split())
        merges[(idx1, idx2)] = idx
        idx += 1
    f.close()

    self.merges = merges
    self.vocab = self._build_vocab(merges)

File: c++ versions/basic.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <numeric>

class BasicTokenizer {
private:
    std::string train_data;
    std::vector<char> chars;
    std::unordered_map<char, int> string_to_index;
    std::unordered_map<int, char> index_to_string;
    std::unordered_map<std::pair<char, char>, int, boost::hash<std::pair<char, char>>> merges;
    std::unordered_map<int, std::vector<char>> vocab;

    std::vector<int> encode(const std::string& text) {
        std::vector<int> encoded;
        for (char c : text) {
            encoded.push_back(string_to_index[c]);
        }
        return encoded;
    }

    std::string decode(const std::vector<int>& integer) {
        std::string decoded;
        for (int i : integer) {
            decoded.push_back(index_to_string[i]);
        }
        return decoded;
    }

    std::unordered_map<std::pair<char, char>, int, boost::hash<std::pair<char, char>>> get_stats(const std::vector<int>& ids) {
        std::unordered_map<std::pair<char, char>, int, boost::hash<std::pair<char, char>>> counts;
        for (size_t i = 0; i < ids.size() - 1; ++i) {
            counts[{ids[i], ids[i + 1]}]++;
        }
        return counts;
    }

    std::vector<int> merge(const std::vector<int>& ids, const std::pair<char, char>& pair, int idx) {
        std::vector<int> new_ids;
        for (size_t i = 0; i < ids.size(); ++i) {
            if (i + 1 < ids.size() && ids[i] == pair.first && ids[i + 1] == pair.second) {
                new_ids.push_back(idx);
                i++;
            } else {
                new_ids.push_back(ids[i]);
            }
        }
        return new_ids;
    }

public:
    BasicTokenizer(const std::string& train_text) : train_data(train_text) {
        chars = std::vector<char>(train_text.begin(), train_text.end());
        std::sort(chars.begin(), chars.end());
        for (size_t i = 0; i < chars.size(); ++i) {
            string_to_index[chars[i]] = i;
            index_to_string[i] = chars[i];
        }
    }

    std::unordered_map<int, std::vector<char>> train_model(int n_merges) {
        std::vector<int> ids = encode(train_data);

        for (int i = 0; i < n_merges; ++i) {
            auto stats = get_stats(ids);
            auto pair = std::max_element(stats.begin(), stats.end(), [](const auto& a, const auto& b) {
                return a.second < b.second;
            })->first;
            int idx = chars.size() + i;
            ids = merge(ids, pair, idx);
            merges[pair] = idx;
        }

        for (const auto& merge_pair : merges) {
            int idx = merge_pair.second;
            vocab[idx] = vocab[merge_pair.first.first];
            vocab[idx].insert(vocab[idx].end(), vocab[merge_pair.first.second].begin(), vocab[merge_pair.first.second].end());
        }

        return vocab;
    }

    std::vector<int> text_encode(const std::string& text) {
        std::vector<int> tokens = encode(text);
        while (true) {
            auto stats = get_stats(tokens);
            auto pair = std::min_element(stats.begin(), stats.end(), [this](const auto& a, const auto& b) {
                return merges[a.first] < merges[b.first];
            })->first;
            if (merges.find(pair) == merges.end()) {
                break;
            }
            int idx = merges[pair];
            tokens = merge(tokens, pair, idx);
        }
        return tokens;
    }

    std::string text_decode(const std::vector<int>& tokens) {
        std::string text;
        for (int token : tokens) {
            text += decode(vocab[token]);
        }
        return text;
    }

    void save_model(const std::string& prefix) {
        std::ofstream model_file(prefix + ".model");
        model_file << "minibpe v1\n";
        for (const auto& merge_pair : merges) {
            model_file << merge_pair.first.first << " " << merge_pair.first.second << "\n";
        }
        model_file.close();

        std::ofstream vocab_file(prefix + ".vocab");
        for (const auto& vocab_pair : vocab) {
            vocab_file << "[" << text_decode({vocab_pair.first}) << "] " << vocab_pair.first << "\n";
        }
        vocab_file.close();

        std::cout << "Model saved successfully!" << std::endl;
    }
};